### PR TITLE
Only lock once in `is_slice_in_psram`, use `usize` and fix a few comparisons

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description

skip-changelog because it's a small internal change only

### CODE REVIEW

File: esp-hal/src/soc/mod.rs [hunk 1]
```diff
@@ -83,36 +97,40 @@ impl self::efuse::Efuse {
...
-    #[cfg(not(psram))]
-    false
+pub(crate) fn is_valid_psram_address(address: usize) -> bool {
+    addr_in_range(address, psram_range())
 }
 
 #[allow(unused)]
 pub(crate) fn is_slice_in_psram<T>(slice: &[T]) -> bool {
-    let start = slice.as_ptr() as u32;
-    let end = start + slice.len() as u32;
```
Comment (Dominaezzz) [inline esp-hal/src/soc/mod.rs:111]: Ah yeah this was subtly wrong. It didn't take into account the size of `T`.

File: esp-hal/src/soc/mod.rs [hunk 2]
```diff
@@ -83,36 +97,40 @@ impl self::efuse::Efuse {
...
     is_valid_ram_address(address) || is_valid_psram_address(address)
 }
+
+fn slice_in_range<T>(slice: &[T], range: Range<usize>) -> bool {
+    let slice = slice.as_ptr_range();
+    let start = slice.start as usize;
+    let end = slice.end as usize;
+    // `end` is >= `start`, so we don't need to check that `end > range.start`
+    // `end` also points one after the last element, so we need to check that it's
+    // strictly less than `range.end`
```
Comment (Dominaezzz) [inline esp-hal/src/soc/mod.rs:130]: Is `range.end` also one after the last valid memory location? I'm failing to convince myself that `<` is more correct than `<=`

File: esp-hal/src/soc/mod.rs (hunk 1 above)
Comment (bugadani) [reply esp-hal/src/soc/mod.rs:111]: I didn't even notice :)

File: esp-hal/src/soc/mod.rs (hunk 2 above)
Comment (Dominaezzz) [reply esp-hal/src/soc/mod.rs:130]: I'm thinking this needs to be something to the effect of `end <= SOC_DRAM_HIGH + 1`.

Comment (bugadani) [reply esp-hal/src/soc/mod.rs:130]: Yeah I looked through the DRAM constants and they are all exclusive, and PSRAM range is set up as start+length which is also exclusive.

That being said, I'm pretty sure I got this wrong (if I declare a slice that takes up all of DRAM, it needs to be contained within the DRAM range, so `end <= range.end` should probably be the correct one here. I'll revisit tomorrow.

Comment (bugadani) [reply esp-hal/src/soc/mod.rs]: Thanks for noticing this.

**Review by Dominaezzz — APPROVED** [review]
LGTM

**Review by bjoernQ — APPROVED** [review]
LGTM

**Pull request merged**